keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,540 @@
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives [optional]](#alternatives-optional)
+<!-- /toc -->
+
+## Summary
+
+Kubernetes natively supports dynamic provisioning for file and block storage but lacks support for object bucket provisioning.
+Without a formal API to describe and manage bucket provisioning, storage providers are left to write their own provisioners including all of the necessary controller code.
+This "wild west" landscape for bucket provisioning is potentially confusing, overly complex, inconsistent, may slow down Kubernetes adoption by object store users, and may hinder application portability.
+
+We propose a common Kubernetes control-plane API for the management of object store bucket lifecycles which has no dependencies on the underlying providers.
+
+## Motivation
+
+By defining a Kubernetes API for object bucket provsioning, we provide a consistent framework for all object store providers and thus enhance the end user's storage experience. This API allows object store vendors to focus on how they wish to provision buckets instead of writing and testing the controller aspects of provisioning.
+
+### Goals
+
++ Define a _control-plane_ object bucket management API thus relieving object store providers from Kubernetes controller details.
++ Minimize technical ramp-up for storage vendors who have already contributed external storage provisioners (file and block).
++ Make bucket consumption similar to file or block consumption using familiar concepts and commands.
++ Use native Kubernetes resources where possible, again, to keep the bucket experience for users and admins similar to existing storage provisioning.
++ Be unopinionated about the underlying object-store and at the same time provide a flexible API such that provisioner specific features can be supported.
++ Ensure bucket consuming pods wait until the target buckets have been created and are accessible. Thus there is no specific order required for when a bucket claim is created vs. when the app pod is run.
++ Present similar user and admin experiences for both _greenfield_ (new) and _brownfield_ (existing) bucket provisioning.
+
+### Non-Goals
+
++ Update the native Kubernetes PVC-PV API to support object buckets.
++ Define a native _data-plane_ object store API.
++ Handle the small percentage of apps that will not be portable due to use of non-compatible object-store features.
+
+## Proposal
+
+Create a bucket provisioning library, similar to the Kubernetes [sig-storage-lib-external-provisioner](https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/master/controller/controller.go) library, which handles all of the controller and Kubernetes resources related to provisioning.
+This library will implement a simple and familiar API via Custom Resources, and define a _contract_ between app developers and provisioners regarding the Kubernetes resources related to bucket provisioning.
```

Comment:
Per above, I think a gRPC API is a better form of contract than a golang-specific library interface.

Reply:
Is the main point here that gRPC has multi-language support?

Reply:
Yes, that was part of my point.  But also the fact that the gRPC API is its own standalone thing, so documentation and version control tend to be handled better than internal library interfaces.